core/base/abstract_factory.hpp

```diff
@@ -0,0 +1,246 @@
+ * `ComponentsType` (which can be constructed on the fly from parameters to its
+ * constructors).
+ * The generate() method is not declared as virtual, as this allows subclasses
+ * to hide the method with a variant that preserves the compile-time type of the
+ * objects. Instead, implementers should override the generate_impl() method,
+ * which is declared virtual.
+ *
+ * Implementers of concrete factories should consider using the
+ * EnableDefaultFactory mixin to obtain default implementations of utility
+ * methods of PolymorphicObject and AbstractFactory.
+ *
+ * @tparam AbstractProductType  the type of products the factory produces
+ * @tparam ComponentsType  the type of components the factory needs to produce
+ *                         the product
+ */
+template <typename AbstractProductType, typename ComponentsType>
+class AbstractFactory
+    : public EnableAbstractPolymorphicObject<
+          AbstractFactory<AbstractProductType, ComponentsType>> {
+public:
+    using abstract_product_type = AbstractProductType;
+    using components_type = ComponentsType;
+
+    /**
+     * Creates a new product from the given components.
+     *
+     * The method will create an ComponentsType object from the arguments of
+     * this method, and pass it to the generate_impl() function which will
+     * create a new AbstractProductType.
+     *
+     * @tparam Args  types of arguments passed to the constructor of
+     *               ComponentsType
+     *
+     * @param args  arguments passed to the constructor of ComponentsType
+     *
+     * @return an instance of AbstractProductType
+     */
+    template <typename... Args>
+    std::unique_ptr<AbstractProductType> generate(Args &&... args) const
+    {
+        return this->generate_impl({std::forward<Args>(args)...});
+    }
+
+protected:
+    /**
+     * Constructs a new factory on the specified executor.
+     *
+     * @param exec  the executor where the factory should be constructed
+     */
+    AbstractFactory(std::shared_ptr<const Executor> exec)
+        : EnableAbstractPolymorphicObject<AbstractFactory>(std::move(exec))
+    {}
+
+    /**
+     * Constructs a new product from the given components.
+     *
+     * @param args  the components from which to create the product
+     *
+     * @return an instance of AbstractProductType
+     */
+    virtual std::unique_ptr<AbstractProductType> generate_impl(
+        ComponentsType args) const = 0;
+};
+
+
+/**
+ * This mixin provides a default implementation of a concrete factory.
+ *
+ * It implements all the methods of AbstractFactory and PolymorphicObject.
+ * Its implementation of the generate_impl() method delegates the creation of
+ * the product by calling the
+ * `ProductType::ProductType(const ConcreteFactory *, const components_type &)`
+ * constructor. The factory also supports parameters by using the
+ * `ParametersType` structure, which is defined by the user.
```

Comment:
It is, I have it in the unit tests [here](https://github.com/ginkgo-project/ginkgo/pull/46/files#diff-740f961bfbc05b2fc2a808feb7af91d0R43).

Would you prefer if I copy the code, or just say which file to look into?

Reply:
I'm unsure here, after checking the three relevant classes don't seem too big and we may want to have the example available in the documentation (not only in the code), so maybe duplicating is a good option? Otherwise `@link` to the file may be useful here, see [doxygen links](https://www.stack.nl/~dimitri/doxygen/manual/autolink.html).

Reply:
If I remember correctly, Tobias mentioned something about stuff like this earlier.

What I see as a problem with copying code is that the code becomes a comment, and we no longer know if it is correct once we change something in the library as it is not compiler/run as part of the unit tests.

Python includes a nice module called [doctest](https://docs.python.org/3.6/library/doctest.html) which solves the problem. Not sure if something similar exists for C++ and doxygen.

For now, maybe it's better to just link to the code that is verified in the unit tests.